Samantha Erica O'Neil
12208566
DATA100 S17

# Zomato Exercise

Zomato is a restaurant aggregator website. The platform provides information, menus and user-reviews of restaurants as well as food delivery options from partner restaurants in select cities. You have been provided 3 datasets:

- zomato.csv contains restaurant information like average_cost_for_two, cuisines
- location.csv contains location of the restaurants like address, city, locality & etc
- ratings.csv contains aggregate rating for each restaurant

The different files are connected using the `id` columns which can be found in the different files. The id column is a unique identifier to represent a restaurant. You can use this to **merge** the different files.

In [ ]:
import pandas as pd
import numpy as np

1. Merge the zomato, location & ratings file into a signle dataframe and store it to the `merged_df` variable **(3pts)**

In [24]:
zomato = pd.read_csv('zomato.csv')
ratings = pd.read_csv('ratings.csv')
location = pd.read_csv('location.csv')
merged_df = pd.merge(zomato, ratings, on='id')
merged_df = pd.merge(merged_df, location, on='id')

print(merged_df)

            id  average_cost_for_two book_url  \
0      6316125                  1000      NaN   
1     18409457                  1300      NaN   
2      6307689                  1600      NaN   
3     18581637                  1000      NaN   
4      6315438                  3000      NaN   
...        ...                   ...      ...   
6825  18635649                  1000      NaN   
6826   6318581                   400      NaN   
6827   6318578                   450      NaN   
6828   6309612                  1000      NaN   
6829   6318582                   600      NaN   

                                        cuisines currency  \
0                                Japanese, Ramen        P   
1                      Korean, Grill, Korean BBQ        P   
2                             American, Filipino        P   
3                                         Korean        P   
4     Seafood, American, Mediterranean, Japanese        P   
...                                          

In [27]:
assert (6830, 35)==merged_df.shape

cols = ['id', 'address', 'city', 'city_id', 'country_id', 'latitude',
       'locality', 'locality_verbose', 'longitude', 'zipcode',
       'aggregate_rating', 'rating_color', 'rating_text', 'votes',
       'average_cost_for_two', 'book_url', 'cuisines', 'currency', 'deeplink',
       'events_url', 'featured_image', 'has_online_delivery',
       'has_table_booking', 'include_bogo_offers', 'is_book_form_web_view',
       'is_delivering_now', 'is_table_reservation_supported',
       'is_zomato_book_res', 'menu_url', 'mezzo_provider', 'name',
       'photos_url', 'price_range', 'thumb', 'url']

assert sorted(merged_df.columns) == sorted(cols)
print(merged_df)

            id  average_cost_for_two book_url  \
0      6316125                  1000      NaN   
1     18409457                  1300      NaN   
2      6307689                  1600      NaN   
3     18581637                  1000      NaN   
4      6315438                  3000      NaN   
...        ...                   ...      ...   
6825  18635649                  1000      NaN   
6826   6318581                   400      NaN   
6827   6318578                   450      NaN   
6828   6309612                  1000      NaN   
6829   6318582                   600      NaN   

                                        cuisines currency  \
0                                Japanese, Ramen        P   
1                      Korean, Grill, Korean BBQ        P   
2                             American, Filipino        P   
3                                         Korean        P   
4     Seafood, American, Mediterranean, Japanese        P   
...                                          

2. Create a function that returns the average aggregate rating for each city. Expected returned value is a Series datatype which contains name of city as index and average aggregate rating as value. **(2pts)**

In [14]:
def exercise_2(merged_df):
  average_ratings = merged_df.groupby('city')['aggregate_rating'].mean()
  return average_ratings

In [15]:
res_series = exercise_2(merged_df)
assert True == np.isclose(3.302075, res_series[res_series.index == 'Pasay City'],
                          rtol=1e-05, atol=1e-08, equal_nan=False)[0]
assert True == np.isclose(1.890741, res_series[res_series.index == 'Valenzuela City'],
                          rtol=1e-05, atol=1e-08, equal_nan=False)[0]
assert True == np.isclose(1.558333, res_series[res_series.index == 'Malabon City'],
                          rtol=1e-05, atol=1e-08, equal_nan=False)[0]

3. Using the apply function of pandas convert the rating_text to a numeric equivalent. You are to introduce a new column in merged_df called `rating_numeric` which is based on the value from rating_text: **(3pts)**

Excellent = 5 <br>
Very Good = 4 <br>
Good = 3 <br>
Average = 2 <br>
Poor = 1 <br>
Not rated = 0 <br>

In [16]:
rating_mapping = {
    'Excellent': 5,
    'Very Good': 4,
    'Good': 3,
    'Average': 2,
    'Poor': 1
}

merged_df['rating_numeric'] = merged_df['rating_text'].apply(lambda x: rating_mapping.get(x, 0))
print(merged_df[['rating_text', 'rating_numeric']])

     rating_text  rating_numeric
0      Excellent               5
1      Excellent               5
2      Very Good               4
3      Very Good               4
4      Excellent               5
...          ...             ...
6825     Average               2
6826     Average               2
6827   Not rated               0
6828        Good               3
6829   Not rated               0

[6830 rows x 2 columns]


In [17]:
assert 868 == merged_df[merged_df.rating_numeric == 0].shape[0]
assert True == ('rating_numeric' in merged_df.columns)

4. Create a function which filters the merged dataframe to only contain rows coming in from a specific city. Expected returned value is a dataframe **(2pts)**

In [18]:
def exercise_4(merged_df, city):
  filtered_df = merged_df[merged_df['city'] == city]
  return filtered_df

In [19]:

assert 'Makati City' == exercise_4(merged_df, 'Makati City').city.unique()[0]
assert (976, 36) == exercise_4(merged_df, 'Makati City').shape

assert 'Mandaluyong City' == exercise_4(merged_df, 'Mandaluyong City').city.unique()[0]
assert (443, 36) == exercise_4(merged_df, 'Mandaluyong City').shape

assert (0, 36) == exercise_4(merged_df, 'Quezon  City').shape

5. Create a function to filter the dataframe to only contain certain columns passed in the function. Expected returned value is a dataframe **(1pt)**

In [20]:
def exercise_5(merged_df, cols_to_filter):
  filtered_df = merged_df[cols_to_filter]
  return filtered_df

In [21]:
cols = ['name', 'cuisines', 'city', 'aggregate_rating']
assert sorted(cols) == sorted(exercise_5(merged_df, cols).columns)
assert (6830, 4) == exercise_5(merged_df, cols).shape

cols = ['average_cost_for_two', 'book_url', 'cuisines', 'currency', 'deeplink',
       'events_url', 'featured_image', 'has_online_delivery',
       'has_table_booking', 'include_bogo_offers', 'is_book_form_web_view']
assert sorted(cols) == sorted(exercise_5(merged_df, cols).columns)
assert (6830, 11) == exercise_5(merged_df, cols).shape


6.  What is the most common restaurant name in the dataset? **(2pts)** Write down your answer and show code used to come up with the answer.

In [28]:
most_common = merged_df['name'].value_counts().idxmax()
print(most_common)

Starbucks


Based on the code result, the most common restaurant name in the dataset is **Starbucks**.

7. Which restaurant has the most expensive average cost for two? **(2pts)** Write down your answer and show code used to come up with the answer.

In [30]:
most_expensive = merged_df['average_cost_for_two'].max()
most_expensive_restaurant = merged_df.loc[merged_df['average_cost_for_two'] == most_expensive, 'name'].values[0]
print(most_expensive_restaurant)

The Tasting Room


First I decided to compute the value of the most expensive cost for two, then I matched the value to a name of a restaurant. The restaurant with the most expensive cost for two is **The Tasting Room**

8. How many restaurants do not have a zipcode? **(1pt)** Write down your answer and show code used to come up with the answer.

In [31]:
no_zipcode = merged_df['zipcode'].isnull().sum()
print(no_zipcode)

6713


There are **6713** restaurants with no zip code

9. How many restaurants in the dataset allow online delivery? **(1pt)** Write down your answer and show code used to come up with the answer.

In [32]:
online_delivery = (merged_df['has_online_delivery'] == 1).sum()
print(online_delivery)

0


There are no restaurants with online delivery.

10. Which locatlity has more expensive food? Poblacion or Salcedo Village? **(3pts)** Write down your answer and show code used to come up with the answer.


In [34]:
poblacion_cost = merged_df[merged_df['locality'] == 'Poblacion']['average_cost_for_two'].mean()
salcedo_cost = merged_df[merged_df['locality'] == 'Salcedo Village']['average_cost_for_two'].mean()
print("Poblacion: " + str(poblacion_cost))
print("Salcedo Village: " + str(salcedo_cost))

Poblacion: 898.2142857142857
Salcedo Village: 989.1025641025641


The locality with the more expensive food, based on the average cost for two is **Salcedo Village with the mean being 989.10**.